server/schedule/placement/rule_list.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing perlissions and
+// limitations under the License.
+
+package placement
+
+import (
+	"bytes"
+	"sort"
+)
+
+type splitPointType int
+
+const (
+	tStart splitPointType = iota
+	tEnd
+)
+
+// splitPoint represents key that exists in rules.
+type splitPoint struct {
+	typ  splitPointType
+	key  []byte
+	rule *Rule
+}
+
+// A rule slice that keeps ascending order.
+type sortedRules struct {
+	rules []*Rule
+}
+
+func (sr *sortedRules) insertRule(rule *Rule) {
+	i := sort.Search(len(sr.rules), func(i int) bool {
+		return compareRule(sr.rules[i], rule) > 0
+	})
+	if i == len(sr.rules) {
+		sr.rules = append(sr.rules, rule)
+		return
+	}
+	sr.rules = append(sr.rules[:i+1], sr.rules[i:]...)
+	sr.rules[i] = rule
+}
+
+func (sr *sortedRules) deleteRule(rule *Rule) {
+	for i, r := range sr.rules {
+		if r.Key() == rule.Key() {
+			sr.rules = append(sr.rules[:i], sr.rules[i+1:]...)
+			return
+		}
+	}
+}
+
+type rangeRules struct {
+	startKey   []byte
+	rules      []*Rule
+	applyRules []*Rule
+}
+
+type ruleList struct {
+	ranges []rangeRules // ranges[i] contains rules apply to (ranges[i].startKey, ranges[i+1].startKey).
+}
+
+func buildRuleList(rules map[[2]string]*Rule) ruleList {
```

Comment:
How about directly calling this function as `updateRuleList`?

Reply:
good idea